x/qgb/keeper/keeper_valset.go

```diff
@@ -94,9 +94,19 @@ func (k Keeper) GetCurrentValset(ctx sdk.Context) (types.Valset, error) {
 
 		p := sdk.NewInt(k.StakingKeeper.GetLastValidatorPower(ctx, val))
 
-		// FIXME @cmwaters fix this by adding a mapping from val address to evm address
-		evmAddrs := gethcommon.BytesToAddress(val)
-		bv := types.BridgeValidator{Power: p.Uint64(), EvmAddress: evmAddrs.String()}
+		evmAddress, exists := k.GetEVMAddress(ctx, val)
+		if !exists {
+			// This should never happen and indicates a bug in the design of
+			// the system (most likely that a hook wasn't called or a migration
+			// for existing validators wasn't conducted). A validator
+			// should always have an associated EVM address. Fortunately we can
+			// safely recover from this by deriving the default again.
+			ctx.Logger().Error("validator does not have an evm address set")
+			evmAddress = types.DefaultEVMAddress(val)
+			k.SetEVMAddress(ctx, val, evmAddress)
```

Comment:
[non-blocking][could be handled in a separate issue][personal opinion]
I guess at this level it's really important to panic if the EVM address doesn't exist because there is a serious issue with state.

The safety recovery system might be misleading in the case that a validator already changed his EVM address from the default one and then them having it being the default again. In that case, we're setting an invalid value in store.

I agree that if this happens the node will not be able to sync because of a different app hash, but it would be better to panic and just fail instead of putting an invalid value and an error log.

Reply:
> The safety recovery system might be misleading in the case that a validator already changed his EVM address from the default one and then them having it being the default again

If a validator has changed their EVM address then it must be in the store. EVM addresses can't be deleted. Once they're in state the validator always has an associated EVM address.

>  In that case, we're setting an invalid value in store.

How are we setting an invalid value?

> I guess at this level it's really important to panic

We can open this discussion (cc @evan-forbes, @staheri14 @rootulp). From my view, in this case we shouldn't actually panic. QGB is there to support an off-chain process. Any implementation bugs should be isolated and shouldn't result in taking down Celestia's primary service which is publishing blobs. The only problem with handling this error is the threat of an EVM address collision (someone set their EVM address to this default value). We can strengthen this by adding some deterministic randomness (like the blob time which is hard for a user to guess/manipulate). Or we can loop until we find a unique address (expensive but not computationally unbounded).

Reply:
Yes I agree, but this case is something that should never be hit. Hitting it means something is really wrong with the store and I think that if a node's store is changed in any way, it's a big red flag, and it can be a signal to also check if the other state is valid or not.

I consider it a general state error and not just QGB related.

Reply:
Makes sense :+1:, let's leave it for now then